persist/interfaces.go

```diff
@@ -0,0 +1,32 @@
+package persist
+
+import (
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	kbucket "github.com/libp2p/go-libp2p-kbucket"
+
+	log "github.com/ipfs/go-log"
+)
+
+var logSnapshot = log.Logger("dht/snapshot")
+var logSeedProposer = log.Logger("dht/seeds-proposer")
+
+// A SeedsProposer proposes a set of eligible peers from a given set of candidates & fallback peers
+// for seeding the RT.
+type SeedsProposer interface {
+	// Propose takes an optional set of candidates from a snapshot (or nil if none could be loaded),
+	// and a set of fallback peers, and it returns a set of eligible peers that can be
+	// used to seed the given routing table.
+	// Returns nil if it has no more proposals.
+	// Note: Seeding a routing table with the eligible peers will work only if the the dht uses a persistent peerstore across restarts.
+	// This is because we can recover metadata for the candidate peers only if the peerstore was/is persistent.
+	Propose(rt *kbucket.RoutingTable, candidates []peer.ID, fallback []peer.ID) []peer.ID
```

Comment:
@aschmahmann

You can take a look at the discussion [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/315#issue-268346676) & [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334531196) that made us arrive at this interface. The basic rationale is:

**1.** The proposer is meant to _propose_ that the RT be moved from state X -> state Y by applying some criteria on the set of peers P provided to it (candidates + fallbacks). It needs to look at the current RT state to determine what state it should go to next. 

For eg: I would want to write a proposer that wants to ensure that the RT have peers belonging to as many different buckets(CPL's) so that the RT is aware of as may "regions" in the kad tree as is possible from the set of peers provided to it.  

Basically, because the proposer exists to propose peers to seed the RT, it does make sense for it know the current state of the RT.

**2.** Hmm, we could add the host to the interface but... I think your idea of doing the host/peer store filtering in the caller(dht_rt_seeder.go) itself is better as I don't see why the caller would want to pass in the candidates/fallbacks that do not exist in the peerstore or why a proposer would want to propose peers that do not exist in the peerstore. 

Wdyt ?

Reply:
> I would want to write a proposer that wants to ensure that the RT has peers belonging to as many different buckets(CPL's) so that the RT is aware of as may "regions" in the kad tree as is possible from the set of peers provided to it. 

Maybe this was always the intention, but that wasn't quite obvious from @raulk's examples of peer selection from https://github.com/libp2p/go-libp2p-kad-dht/pull/315#issue-268346676

> we can pick candidates randomly, or we can choose based on latency, geographical location (e.g. to favour proximity), etc.

@raulk are we all on the same page that the `Proposer` may utilize the state of the routing table as part of the metric in determining which peers to propose and the ordering of those peers?

I've added some thoughts below https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r355873456 taking this possibility into account.